### PR TITLE
Refactor: Migrate to 2.0-style security policies

### LINKED ISSUE #11211: Refactor Warehouse's Auth{N,Z} to use Pyramid 2.0-style security policies
Pyramid 2.x introduces a [refactored AuthN/AuthZ system](https://docs.pylonsproject.org/projects/pyramid/en/latest/whatsnew-2.0.html#upgrading-auth-20), built around "security policies." These changes combine authentication and authorization into a single policy, and produce an "identity" that's more flexible than the concept of an authenticated user.

Why does Warehouse need this? The most immediate reason is that it simplifies the planned approach for #10970, which involves minting API tokens that aren't tied to any particular user. Because they aren't associated with a user identity, these tokens don't fit into the Pyramid 1.0 authentication scheme.

Here are some open things that need to be considered as part of this refactor:

* How are we going to handle multiple authentication policies? The current codebase uses [`pyramid_multiauth`](https://github.com/mozilla-services/pyramid_multiauth) to dispatch to multiple underlying providers, but that library doesn't support Pyramid 2.0-style policies yet. Tracking: https://github.com/mozilla-services/pyramid_multiauth/issues/28
* How much/how invasive will the `request.user` refactor be? That attribute currently pervades the codebase, and is currently assumed to be a `User` model (AFAICT). Should we remove that entirely, or keep it around as an `Optional[User]`?

Pinging @ewdurbin and @di for thoughts.

### LINKED ISSUE #11252: Don't flash warning on session expiration for token-based sessions
As raised in https://github.com/pypa/warehouse/pull/11218#discussion_r856926749, currently token-based sessions will get a flash message when the session expires, which is unnecessary.

### PR DESCRIPTION
**Pull request by woodruffw**
~~Very WIP; pushing up for visibility.~~

Closes #11211.

Closes #11252.

### CODE REVIEW

**Comment by woodruffw** [comment]
This currently partially works (I can log in and navigate around), but permissions are currently broken for reasons I haven't fully debugged (trying to manage projects returns a `403`).

**Comment by woodruffw** [comment]
This needs a lot of testing, but the basic structure is here:

* Each of session, basic, and macaroon-based authentication uses a "security policy" rather than 1.0-style AuthN APIs.
* We provide a new `MultiSecurityPolicy` that wraps multiple security policies in a manner similar to the now-removed `pyramid-multiauth`. One significant deviation: `MultiSecurityPolicy` also handles AuthZ, but it doesn't use each policy's separate AuthZ -- instead it keeps a singular AuthZ policy internal to itself, since Warehouse only has one.
* Our "identity" structure is now a dictionary with the following format:
    
    ```py
    { "entity": entity, "principals": principals }
    ```

...where `entity` can be `User` model or any other DB model suitable for a particular AuthN/AuthZ flow, and `principals` is the list of relevant security principles (previously a dedicated part of Pyramid's AuthZ APIs).

File: warehouse/utils/security_policy.py
Comment (woodruffw) [inline warehouse/utils/security_policy.py:24]: N.B. This is currently a generic `identity` for when policies always boil down to user identities. I'll probably need to make it less generic or remote it entirely when we allow for project identities.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [inline warehouse/utils/security_policy.py:90]: Flagging this for review: I cargo-culted this from the Pyramid cookbook. I don't _think_ it causes access issues since `Everyone` means "everyone, all the time" and as such it's sort of a no-op check against most routes, but I'm not certain.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [inline warehouse/utils/security_policy.py:62]: Also flagging this for design review: Warehouse only has one real AuthZ policy (well, multiple, but they're composed as a nested set), so I wasn't sure if it made sense for each security policy to keep its own copy of the AuthZ policy. Currently they don't and only the top-level `MultiSecurityPolicy` performs the `permits` check, but that's a bit of an abstraction violation.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [inline warehouse/utils/security_policy.py:39]: See https://github.com/pypa/warehouse/pull/11218/files#r854535588 for design review on this.

File: warehouse/utils/security_policy.py
Comment (di) [inline warehouse/utils/security_policy.py:36]: I think I'd prefer this to just return the ORM object that corresponds to the identity (here, the user). The computation of the principals can happen when we need it in `permits` instead, we don't need to pass it around, and we don't need principals available everywhere `request.identity` would be available.

I also think the `identity` function for `SecurityPolicy` should probably be `NotImplemented` and the subclass should be responsible for generating the corresponding identity instead. This means that `SessionSecurityPolicy` and `BasicAuthSecurityPolicy` could always return `Users`, but we could eventually update `MacaroonSecurityPolicy` to return other identities.

Since for now, they would all share the same logic for computing a user from a request, perhaps we turn this into a standalone function or mixin or something instead.

File: warehouse/utils/security_policy.py
Comment (di) [reply warehouse/utils/security_policy.py:62]: I think this is OK, at least until we decide we want to overhaul our use of `principals` entirely.

(I think eventually `permits` would just determine outcomes based on the relationship (e.g. `Roles`) between `identity` and `context`, but that would be a big change that we don't need for this).

File: warehouse/utils/security_policy.py
Comment (di) [inline warehouse/utils/security_policy.py:26]: This is also going to have to change once `MacaroonSecurityPolicy` is not always returning a `User`, so probably better to delegate this to the individual policies instead.

File: warehouse/accounts/__init__.py
Comment (woodruffw) [reply warehouse/accounts/__init__.py:218]: The only reason was "minimize code churn" 🙂 -- I could move them into their respective `security_policy.py` files and eliminate the callbacks if you don't mind a slightly bigger diff.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py:36]: Yeah, this is another case where I cargo-culted it from Pyramid's porting guide: they recommended having the identity contain the principles, but I agree that there's no reason for it to. I'll refactor so that it's just the ORM model.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py:90]: None that I can think of, they should be functionally equivalent. Will drop.

File: warehouse/accounts/__init__.py
Comment (di) [reply warehouse/accounts/__init__.py:218]: I think that'd be preferable, would make it slightly easier to follow what's calling what.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py:36]: Oh, I remember why I did this now -- our AuthZ policies are still using the old APIs, which require the principals to be passed in explicitly. I couldn't figure out how else to get them into those APIs without stashing them on the `identity`, since the only context available is `request`, `context` (already set), and `permissions` (already set?)

File: warehouse/utils/security_policy.py
Comment (di) [reply warehouse/utils/security_policy.py:90]: Not necessarily asking to drop it, just don't understand what it's purpose is. Maybe we should be doing it?

File: warehouse/utils/security_policy.py
Comment (di) [reply warehouse/utils/security_policy.py:36]: I think they can be computed in `permits` right before we pass them to the AuthZ policy, no?

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py:90]: If I understand how Pyramid does ACLs correctly, it's essentially the "no-op" principal -- every identity belongs to `Everyone`, including "no-one", so there's no difference between it being present and not.

That's per the docs:

> [](https://docs.pylonsproject.org/projects/pyramid/en/latest/api/authorization.html#pyramid.authorization.Everyone)Everyone
The special principal id named Everyone. This principal id is granted to all requests. Its actual value is the string 'system.Everyone'.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py:36]: Yeah, I think you're right (I was overthinking it). `permits` can use `request.identity` as long as it's present, so we can do anything that the policy's `identity` would do anyways. I'll refactor!

Comment (woodruffw) [reply warehouse/utils/security_policy.py]: [0b788d9](https://github.com/pypa/warehouse/pull/11218/commits/0b788d91af438990ebd4961eb070c0ce31af468e)

File: warehouse/accounts/__init__.py
Comment (woodruffw) [reply warehouse/accounts/__init__.py]: Done. A lot of this boiled down to principal collection as well, so I was able to simply remove most of it and put the remainder in a `principals_for_authenticated_user` helper.

File: warehouse/accounts/security_policy.py
Comment (woodruffw) [inline warehouse/accounts/security_policy.py:120]: Flagging for review: this is a Pyramid API so I'm assuming it's correct, but I'm a little bit confused by *how* a session gets authenticated. For basic auth or macaroons the logic is embedded in the `identity` body, but here we just seem to grab a `userid` from thin air. Is this helper using the (pre-validated?) session machinery internally?

File: warehouse/macaroons/security_policy.py
Comment (woodruffw) [inline warehouse/macaroons/security_policy.py:82]: Flagging this for review as well: we didn't actually have this before, which makes me think that our authn/z for macaroons might have been a little too permissive (we were probably allowing anything with `context=Project` that matched the token's actual permissions).

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py]: Removed it and nothing broke.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py]: Removed this and fully delegated.

File: warehouse/macaroons/security_policy.py
Comment (di) [reply warehouse/macaroons/security_policy.py:82]: Yes, this is known: https://github.com/pypa/warehouse/issues/7266

Comment (di) [reply warehouse/macaroons/security_policy.py:82]: (I'm also not convinced that we need this, at least in this PR).

File: warehouse/accounts/security_policy.py
Comment (dstufft) [reply warehouse/accounts/security_policy.py:120]: There's two APIs here:

1. ``unauthenticated_userid`` which is intended to pull the userid out of the request.
2. ``authenticated_userid`` which is intended to consume ``unauthenticated_userid`` and ensure that it matches what is in the data store.

For sessions, the data store is the session key, and when ``remember`` is called, the user id is stored in the user's Session, then ``authenticated_userid`` pulls that userid out of the session. Since session storage doesn't have passwords or anything like that and relies on the session id being randomly generated and secret, it just accepts whatever is in the session storage as the authenticated user id.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py]: Removed.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py]: Removed.

File: warehouse/macaroons/security_policy.py
Comment (woodruffw) [reply warehouse/macaroons/security_policy.py:82]: Totally missed that check! I can remove this route check then, since it's a little out of place in the AuthN section.

File: warehouse/accounts/security_policy.py
Comment (woodruffw) [reply warehouse/accounts/security_policy.py:120]: Makes sense, thanks for the explanation. For some reason I forgot that session tokens are bearers and that that's the trust relationship.

File: warehouse/macaroons/security_policy.py
Comment (woodruffw) [reply warehouse/macaroons/security_policy.py]: [0bc2083](https://github.com/pypa/warehouse/pull/11218/commits/0bc208360ef2f1f94146e35396e4ef981a4b77c1)

File: warehouse/macaroons/security_policy.py
Comment (ewjoachim) [inline warehouse/macaroons/security_policy.py:95]: `macaroon_service.find_userid` extracts the macaroon ID from the macaroon and then fetches the macaroon with the ORM joining with the user table, then extracts the user id from the user, and discards both the macaroon instance and the user instance. And then you make a second request to get the user from its id.

You'd gain a DB request by doing 
```python
        user = macaroon_service.find_from_raw(macaroon).user
```

(also, I wonder why there's no exception handling here but there wasn't before either)

File: warehouse/utils/security_policy.py
Comment (di) [reply warehouse/utils/security_policy.py:42]: I believe you're right, but since that's not a bug introduced by this PR I've made a new issue to capture it: https://github.com/pypa/warehouse/issues/11252

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py:120]: Yeah, it's extremely confusing -- it's because we're really wrapping the old-style AuthZ APIs under the hood here. I'll add a comment that explains the behavior a bit.

File: warehouse/accounts/security_policy.py
Comment (di) [reply warehouse/accounts/security_policy.py:52]: Good point. I'd be in favor of making this change after this PR is merged.

File: warehouse/accounts/security_policy.py
Comment (di) [reply warehouse/accounts/security_policy.py:72]: Ditto here.

File: warehouse/accounts/__init__.py
Comment (woodruffw) [reply warehouse/accounts/__init__.py:81]: If the goal is to completely switch over to 2.0-style policies, we should probably eventually fold the big callback tower into a single AuthZ policy "helper", since that's the pattern that Pyramid itself uses for e.g. basic and session-based auth 🙂

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py:111]: > Writing the id in the principals seems slightly dangerous to me because there could be a confusion if a user and a project had the same id. I know it's very very unlikely due to both ids being UUIDs, but I don't think it's a good idea to receive an ID and to have to guess what model it belongs to.

I might be misunderstanding what you mean, but we should never actually have to guess the model -- `request.identity` will always be the correct ORM model, so the `id` here is solely for permission matching.

(That being said, your suggestion re: `f"user:{user.id}"` sounds reasonable to me, if we do decide that raw IDs are ambiguous.)

File: warehouse/macaroons/security_policy.py
Comment (di) [reply warehouse/macaroons/security_policy.py:95]: This isn't going to be around very long, as we'll need a way to get either the User or the Project from the macaroon, but I'd be fine with making this change for now to make this slightly more efficient.

File: warehouse/accounts/security_policy.py
Comment (woodruffw) [reply warehouse/accounts/security_policy.py:52]: Yep, agreed. I'll create a rollup issue tracking these kinds of post-merge refactors.

File: warehouse/utils/security_policy.py
Comment (di) [reply warehouse/utils/security_policy.py:111]: I think what @ewjoachim is that it's possible (but unlikely) for an `id` to be shared by both a `User` and a `Project`, which I think is correct.

I'm not seeing why we're doing this though, were we doing it before and I'm just missing it? Or is this something taken from the new 2.0-style examples? (I'm not sure it's necessary).

File: warehouse/accounts/security_policy.py
Comment (woodruffw) [reply warehouse/accounts/security_policy.py:170]: I cribbed this directly from the previous (Pyramid 1.0-style) authn policy: we were already implicitly returning this header, so I figured we should continue to do so on the (exceedingly unlikely) chance that someone is actually expecting it.

> I'm especially doubtful given that our MultiAuthSecurityPolicy calls remember on every subpolicies when we call its remember method, so as far as I can tell, this would result in this header be sent back to the browser when we log in ? And I guess the reason the browser doesn't display an auth modal is because it's not accompanied by a 401 Unauthenticated ?

Yep, correct. The previous (`pyramid_multiauth`-based) implementation also did this, mind you: https://github.com/mozilla-services/pyramid_multiauth/blob/e3d2f5d75201b05daefd0c589f44fb55fdf8e877/pyramid_multiauth/__init__.py#L137-L146

(I think it's pretty ugly that we always return all `remember` and `forget` headers, but I'm not sure there's a straightforward way to do that while composing AuthN policies in the way we are.)

File: warehouse/accounts/security_policy.py
Comment (woodruffw) [reply warehouse/accounts/security_policy.py:43]: I think you're right about this. I'll include it in the refactor rollup.

Comment (woodruffw) [reply warehouse/accounts/security_policy.py:43]: https://github.com/pypa/warehouse/issues/11253

File: warehouse/accounts/security_policy.py
Comment (woodruffw) [reply warehouse/accounts/security_policy.py:72]: https://github.com/pypa/warehouse/issues/11253

File: warehouse/accounts/security_policy.py
Comment (woodruffw) [reply warehouse/accounts/security_policy.py:52]: https://github.com/pypa/warehouse/issues/11253

File: warehouse/macaroons/security_policy.py
Comment (woodruffw) [reply warehouse/macaroons/security_policy.py]: 5cdb53a6

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py:111]: Yeah, it's another thing I cargo-culted from the upgrade guide. I can probably remove it entirely with nothing breaking, will experiment now.

Comment (woodruffw) [reply warehouse/utils/security_policy.py]: Hmm. I tried to remove this and things *appear* to work, but package upload with HTTP basic auth always 403s without it. Any thoughts?

Comment (dstufft) [reply warehouse/utils/security_policy.py:110]: https://github.com/pypa/warehouse/blob/main/warehouse/packaging/models.py#L217-L236

Comment (woodruffw) [reply warehouse/utils/security_policy.py:110]: Thanks @dstufft! That's an easy fix.

Comment (di) [reply warehouse/utils/security_policy.py:110]: Yep, that's it. So I think @ewjoachim's idea to prefix this with `user:` makes sense.

Comment (woodruffw) [reply warehouse/utils/security_policy.py:110]: @dstufft do you happen to know how the original code was placing the user ID in the principals list? I'm having trouble finding it.

Comment (dstufft) [reply warehouse/utils/security_policy.py:110]: https://github.com/pypa/warehouse/blob/main/warehouse/accounts/__init__.py#L66-L92

Comment (dstufft) [reply warehouse/utils/security_policy.py:110]: I don't remember why I used strings there instead of ORM objects, there might have been a reason, but it might also work just fine to use objects. I think probably the examples all used strings, and I didn't feel like digging into it to figure out if not-strings would introduce some horrible security bug.

Comment (woodruffw) [reply warehouse/utils/security_policy.py:110]: I basically preserved that function as `_principals_for_authenticated_user`, but I don't actually see the user's ID getting added to the principal set there. Unless I'm missing something?

Comment (dstufft) [reply warehouse/utils/security_policy.py]: Oh right. Sorry, the ``request.effective_principals`` in Pyramid pre 2.0 (no idea what happens after that) gets populated by the authentication policy's ``effective_principals`` method. Both the Session and the Basic auth authN policy that ships (shipped?) with Pyramid used the ``CallbackAuthenticationPolicy``, which looks like:

https://github.com/Pylons/pyramid/blob/1.9-branch/pyramid/authentication.py#L107-L184

Which puts the special "Everyone" ID, the string of the user id, and then anything that comes from the "groups" callback, which is our ``_authenticate`` method.

It looks like the Macaroon policy also uses the ``CallbackAuthenticationPolicy``, so it will also do that.

Comment (woodruffw) [reply warehouse/utils/security_policy.py]: Mystery solved! 😄 

Yeah, we've gotten rid of the callbacks entirely (since they're not part of the 2.0 APIs), so now everything is concentrated into that one `permits` method that dispatches into the underlying stack of AuthZ policies.

Comment (dstufft) [reply warehouse/utils/security_policy.py]: Assuming something like this still exists in Pyramid 2.0, we could just ignore the fact there is a string-ified userid in the principals, and make the ``_authenticate`` method also return a ``f"{model}:{id}"`` prinicpal, and actually use that in our ACLs OR we can override some methods to make our own ``CallbackAuthenticationPolicy`` that has the behavior we want.

Comment (dstufft) [reply warehouse/utils/security_policy.py]: Ah cool, that's easier then.

**Comment by woodruffw** [comment]
I need to do a final round of local testing, but this should be nearly good to go.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [inline warehouse/utils/security_policy.py:37]: This was part of the original code so I haven't moved it, but it's oddly placed here -- it's really part of AuthN, not the AuthZ work we're doing here. Any objections to me moving it into the `identity`/AuthN phase instead?

Comment (woodruffw) [reply warehouse/utils/security_policy.py]: Done with [29b40f9](https://github.com/pypa/warehouse/pull/11218/commits/29b40f969b7f5191b89189f82b061854295226cf)

**Review by ewjoachim — APPROVED** [review]

**Review by di — APPROVED** [review]
This looks great. Leaving this open for a little while to give @ewdurbin and @dstufft a chance to review if they are able.

**Comment by woodruffw** [comment]
This is the main blocker for #11272, which will "finish" the core functionality here (allowing a user to go from a JWT to an ephemeral API token).

File: warehouse/utils/security_policy.py
Comment (di) [inline warehouse/utils/security_policy.py:78]: This is fine for now but is going to be problematic once we have non-user identities, since these won't have a prefix like the principals would. We might want to address that here in advance, e.g. by asserting that the identity is a user  and otherwise returning `None` or something similar.

**Review by di — APPROVED** [review]
This is ready to merge once this last comment is addressed.

File: warehouse/utils/security_policy.py
Comment (woodruffw) [reply warehouse/utils/security_policy.py:78]: Yeah, I'll make it so that only `User` identities return an ID here. That'll be consistent with our expected behavior, i.e. that `request.user` will only ever be populated with `User` models.

Comment (woodruffw) [reply warehouse/utils/security_policy.py]: [ec2c563](https://github.com/pypa/warehouse/pull/11218/commits/ec2c563d4003556435664911a23364c7bac20197)

**Review by di — APPROVED** [review]

**Pull request merged**
